src/cljam/algo/pileup.clj

```diff
@@ -142,6 +115,156 @@
          (partial filterv)
          (update p 1))))
 
+(defn- merge-corrected-quals
+  "Merge corrected quals with the uncorrected part."
+  [^SAMAlignment aln ^long correct-start corrected-quals]
+  (let [quals (:quals-at-ref aln)
+        start (.pos aln)
+        len   (count corrected-quals)]
+    (if (<= start correct-start)
+      (into (into (vec (take (- correct-start start) quals)) corrected-quals)
+            (drop (+ (- correct-start start) len) quals))
+      (into corrected-quals (drop len quals)))))
+
+(defn- correct-pair-quals
+  "Correct quals of a pair. Returns a map with corrected quals."
+  [^SAMAlignment a1 ^SAMAlignment a2]
+  (if-not (and (pos? (.pnext a1))
+               (<= (max (.pnext a1) (.pos a2)) (.end a1)))
+    {}
+    (let [tlen1 (.tlen a1)
+          tlen2 (.tlen a2)
+          quals1 (:quals-at-ref a1)
+          quals2 (:quals-at-ref a2)
+          seqs1  (:seqs-at-ref a1)
+          seqs2  (:seqs-at-ref a2)
+          correct-start (max (.pnext a1) (.pos a2))
+          new-quals (for [pos (range correct-start
+                                     (inc (min (.end a1) (.end a2))))]
+                      (let [relative-pos1 (- pos (.pos a1))
+                            relative-pos2 (- pos (.pos a2))
+                            q1 (quals1 relative-pos1)
+                            q2 (quals2 relative-pos2)
+                            [b1 _] (seqs1 relative-pos1)
+                            [b2 _] (seqs2 relative-pos2)]
+                         (if (= b1 b2)
+                           [(min 200 (+ q1 q2)) 0]
+                           (if (<= q2 q1)
+                             [(int (* 0.8 q1)) 0]
+                             [0 (int (* 0.8 q2))]))))
+          [new1 new2] (apply map vector new-quals)]
+      {(keyword (str tlen1))
+       (merge-corrected-quals a1 correct-start new1)
+       (keyword (str tlen2))
+       (merge-corrected-quals a2 correct-start new2)})))
+
+(defn- make-corrected-quals-map
+  "Make a map which has corrected quals of all overlapping pairs."
+  [alns]
+  (->> alns
+       (group-by (fn [x] (.qname ^SAMAlignment x)))
+       (into {} (map
+                 (fn [[qname xs]]
+                   (if (<= (count xs) 1)
+                     {}
+                     {(keyword qname) (apply correct-pair-quals xs)}))))))
+
+(defn- correct-quals-at-ref
+  "Returns an alignment with corrected quals by looking up quals
+  in the corrected quals map."
+  [corrected-map ^SAMAlignment aln]
+  (let [qname (.qname aln)
+        tlen  (.tlen aln)]
+    (if (zero? tlen)
+      aln
+      (if-let [new-quals (get-in corrected-map
+                                 [(keyword qname) (keyword (str tlen))])]
+        (assoc aln :quals-at-ref new-quals)
+        aln))))
+
+(defn- get-max-pnext
+  "Get the maximum pnext of alignments."
+  [alns]
+  (->> alns
+       (map #(.pnext ^SAMAlignment %))
```

Comment:
You should check first whether or not `(= (.rnext %) "=")` here, or you could read in a read from a wrong position, although it's relatively rare that you find those pair whose one read has `.rnext` different from the other read's because `basic-mpileup-pred` usually drops such a "poorly aligned" read.

Reply:
Thanks!
I've inserted the condition.